### DISCUSSION TITLE
serialization/deserialization for LanguageServer?

### DISCUSSION THREAD

**Discussion by onriv** [redhat-developer/lsp4ij#422] [Q&A] [answered]
Hi during the implementation of customized `LanguageServer`, some types and messages return from the lsp server end require some specific serialization/deserialization logic (like deserializing to different subclasses)

Any way to do this? Thanks!

**Comment by CppCXY** [comment]
If you're interested in how to handle messages about the language server, you can refer to my project, although it is implemented in C#. see: https://github.com/CppCXY/LanguageServer.Framework

**Comment by onriv** [comment]
@CppCXY  Thanks! Checking the source code adding some TypeAdapter to the gson object [here](https://github.com/redhat-developer/lsp4ij/blob/533f5b215acbcbe7fba14e9626f96ed882a46cf3/src/main/java/com/redhat/devtools/lsp4ij/JSONUtils.java#L36) may be enough for my use case. Or after deserializing the resp using the default gson object, I serialize it again and then desrialize to my target class.

**Comment by onriv** [comment]
The use case here is for deserializing some heterogenous json msg from the server like `["a":1, "b":{"c":2}]` consisting different types in a list. Gson requires customizing some TypeAdapter for it and I don't know if Lsp4ij has such api or not.

**Comment by onriv** [comment, marked as answer]
I find https://github.com/eclipse-lsp4j/lsp4j/issues/768 quite similar to my case too. I will give this a try later

**Comment by angelozerr** [comment]
@onriv where do you need to convert messages? From data of lsp command? From data of lsp code action? Other?

**Comment by onriv** [comment]
@angelozerr I am trying to add customizing `LanguageServer` in the factory
```
    @Override //If you need to expose a custom server API
    public @NotNull Class<? extends LanguageServer> getServerInterface() {
        return MyCustomServerAPI.class;
    }
```
and the resp of the api in `MyCustomServerAPI` requires some customized deserialization. I find that Gson's annotation [JsonAdapter](https://github.com/google/gson/blob/main/gson/src/main/java/com/google/gson/annotations/JsonAdapter.java) seems enough for this and it seems like more related to lsp4j
